keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -73,279 +75,320 @@
-    |   conditions:                        |
-    |   - type: BucketReady                |
-    |     status: False                    |
-    |--------------------------------------|
-```
-
-###### 3. COSI calls appropriate driver to create Bucket
-
-The COSI sidecar, which runs alongside each of the drivers, listens for Bucket events. All but the sidecar with the specified driver will ignore the Bucket object. Only the appropriate sidecar will make a gRPC request to create a Bucket in the OSP.
-
-More information about COSI gRPC API is [here](#cosi-grpc-api)
-
-```
-    COSI Driver (s3.amazonaws.com)
-    |------------------------------------------|
-    | grpc DriverCreateBucket({                |
-    |      "name": "bcl-$uuid",                |
-    |      "protocols": ["s3"],                |
-    |      "parameters": {                     |
-    |         "key": "value"                   |
-    |      }                                   |
-    | })                                       |
-    |------------------------------------------|
-```
-Once the Bucket is successfully created in the OSP, and a successful status is retrieved from the gRPC request, sidecar sets `BucketReady`condition to `True`. At this point, the Bucket is ready to be utilized by workloads.
-
-## Generating Access Credentials for Buckets
+1.  **Admin creates BucketClass, User creates BucketClaim**
+
+    The BucketClass represents a set of common properties shared by multiple buckets. It is used to specify the driver for creating Buckets, and also for configuring driver-specific parameters. More information about BucketClass is [here](#bucketclass)
+
+    ```yaml
+    apiVersion: objectstorage.k8s.io/v1alpha2
+    kind: BucketClass
+    metadata:
+      name: silver
+    spec:
+      deletionPolicy: delete
+      driverName: s3.amazonaws.com
+      parameters:
+        key: value
+    ```
+
+    The BucketClaim is a claim to create a new Bucket. This resource can be used to specify the bucketClassName, which in-turn includes configuration pertinent to bucket creation. More information about BucketClaim is available [here](#bucketclaim)
+
+    ```yaml
+    apiVersion: objectstorage.k8s.io/v1alpha2
+    kind: BucketClaim
+    metadata:
+      name: my-bucket
+      namespace: my-namespace
+    spec:
+      bucketClassName: silver
+      protocols:
+        - S3
+      # existingBucketName: name of pre-existing Bucket object
+    ```
+
+2.  **COSI creates an intermediate Bucket object**
+
+    The ControllerManager creates a Bucket object by copying fields over from both the BucketClaim and BucketClass. This step can only proceed if the BucketClass exists. This Bucket object is an intermediate object, with its status condition `BucketReady` to false. This indicates that the Bucket is yet to be created in the OSP.
+
+    <!-- TODO: make sure there is a mechanism to ensure that the user creating the BucketClaim has
+              access to the bucketclass -->
```

Comment:
@xing-yang I am curious how other projects like VolumeSnapshotter handle situations like this.

How can the COSI controller check that the user (or namespace, or serviceaccount) has proper permissions to read/use the BucketClass?